Suppress the ipv6 no-suitable-address huge error

naiming-zededa opened the pull request:
Signed-off-by: Naiming Shen <naiming@zededa.com>

 Fix the long download error message due to 'no suitable address found' on IPv6
 - for http download, reduce multiple error lines into one, and log will show
   the message, error reason string only has the short 'skip no suitable address' string
   even for multiple ipv6 addresses and interfaces
 - for azure, s3, and docker only handles at syncop.go and resolveconfig.go routines
   since they don't retry multiple times inside
 - tested for http, azure, s3 and docker mechanisms

eriknordmark commented on pkg/pillar/cmd/downloader/resolveconfig.go:
```diff
@@ -275,7 +276,11 @@ func resolveTagsToHash(ctx *downloaderContext, rc types.ResolveConfig,
 				break
 			}
-			errStr = errStr + "\n" + err.Error()
+			if strings.Contains(err.Error(), logutils.NoSuitableAddrStr) {
```
This is not the go way to look at errors; an error isn't a string just because it has a string rendition. Folks have written about better ways. See e.g., https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

I don't know the best way to do this using the old net package, but a slightly better is to traverse the error hiearchy. When looking for an example I found isNoSuitableAddress() in send.go.
Note that it still ends at a string compare because the standard net package does not export the variable (or constant) which has that string rendition.

naiming-zededa replied:
actually i used the isNoSuitableAddress() for http scheme, but it breaks in the e2 which trying to check the AddrError and to be not ok. I tried to find if there is any other error scheme under e1.Err, but could not find. and for azure, aws, docker, those errors don't even fit into url.Error(), but I didn't try to explore all the error available

eriknordmark replied:
Seems like too many go packages don't follow the golang recommendations on how to generate and handle errors, Sigh,

eriknordmark commented on pkg/pillar/cmd/downloader/syncop.go:
```diff
@@ -214,8 +216,18 @@ func handleSyncOp(ctx *downloaderContext, key string,
+			// short string "skip no suitable address found" after the real error
 			sourceFailureError(ipSrc.String(), ifname, metricsURL, err)
-			errStr = errStr + "\n" + err.Error()
+			if strings.Contains(err.Error(), logutils.NoSuitableAddrStr) {
```
Ditto

eriknordmark commented on pkg/pillar/cmd/downloader/syncop.go:
```diff
@@ -175,6 +176,7 @@ func handleSyncOp(ctx *downloaderContext, key string,
 	}
 
 	// Loop through all interfaces until a success
+	var gotNoSuitableAddress bool
```
Is there a behavioral difference between this code using a boolean and the send.go code using a counter for the same purpose? Can we make them use the same approach reduce the number of variants the reader needs to understand?

naiming-zededa replied:
i think the difference is for download, its not just a error spit out in the log, it forms an error-String to be used to publish for this download failure reason. I want to mainly reduce the size of that string, thus for multiple interfaces, if we got the same error, we just form one error-string to be used for publish, but the logging is not suppressed on this, similar to the send.go

eriknordmark reviewed (changes requested):
If I get an error using IPv4 (a timeout, 404, whatever) and then the code tries IPv6 but has no addresses, will this completely supress the IPv6 part of the error?

The no suitable address error only makes sense when there was no other more relevant error, a bit like the case of the explicit check for no IP addresses at all.

naiming-zededa commented:
patch updated:
- removed the 'skip no suitable address found' at the errStr for display
- improved the handling of errors at various downloading schemes
- added more comment in the code

naiming-zededa commented:
yetus still has error on:
ERROR: Failed to write github status. Token expired or missing repo:status write?

eriknordmark commented on pkg/pillar/utils/logging/log.go:
```diff
@@ -32,3 +34,33 @@ func GetMyStack() string {
 	return output
 }
+
+// NoSuitableAddrStr - the string to be used for checking the error message
```
How about adding a comment that this should really be exported by the standard net package but isn't so we end up with a duplicate definition here.

eriknordmark commented on pkg/pillar/utils/logging/log.go:
```diff
@@ -32,3 +34,33 @@ func GetMyStack() string {
+const NoSuitableAddrStr string = "no suitable address found"
+
+// NoSuitableAddrErr - the function to check 'no suitable address' for http mainly
+func NoSuitableAddrErr(err error) bool {
```
Nit, but shouldn't this function be called IsNoSuitableAddrErr as in IsEOF, IsTemporary and other predicates on errors?

eriknordmark reviewed (approved):
Kicking off the tests in parallel with the comment and naming nits.

naiming-zededa commented:
updated per comment.

eriknordmark commented on pkg/pillar/cmd/loguploader/loguploader.go:
```diff
@@ -718,6 +718,8 @@ func getTimeNumber(isApp bool, fName string) (bool, int) {
 	return false, 0
 }
 
+var hackurlCnt int
```
This is unused. Delete?

naiming-zededa replied:
Ops, done.

eriknordmark reviewed (approved):
Ignoring the spurious yetus error.

giggsoff commented on libs/zedUpload/httputil/http.go:
```diff
@@ -18,6 +18,7 @@ import (
 	"strings"
 	"time"
 
+	logutils "github.com/lf-edge/eve/pkg/pillar/utils/logging"
```
Seems it requires go.mod update to include `github.com/lf-edge/eve/pkg/pillar`. If I correctly understood, this library is used by cloud side.

naiming-zededa replied:
i did 'make proto-vendor'. do you see any compiling issue @giggsoff ?

giggsoff replied:
I can see problems with `go get github.com/lf-edge/eve/libs/zedUpload`:
```
...
go get: github.com/lf-edge/eve/pkg/pillar@v0.0.0-20210922101739-3840a435ad43 requires
        github.com/lf-edge/eve/api/go@v0.0.0-00010101000000-000000000000: invalid version: unknown revision 000000000000
```

Seems, it comes with our dummy version in pillar/go.mod.
